ENH: ufuncs can take a tuple of arrays as 'out' kwarg

jaimefrio opened the pull request:
Closes #4752

jaimefrio committed:
ENH: ufuncs can take a tuple of arrays as 'out' kwarg

Closes #4752

jaimefrio commented:
As discussed in #4752 this is currently failing the tests, because these expect ufuncs with multiple returns to accept a single array in the `out` kwarg to be used for the first of the returns.

**Old behavior**
* Inputs are the `nin` first positional arguments
* Outputs are the next `nout` positional arguments.
* If less than `nout` are provided, they default to `None` (i.e. allocated by iterator)
* If no output array has been provided positionally, then the first output array can be provided with the `out` keyword argument.

**New behavior**
* Positional inputs are unchanged
* If no output array has been provided positionally, then:
  * the `out` keyword argument can be a tuple of exactly `nout` items, which must all either be arrays or `None`.
  * if `nout == 1`, then a single array or `None` can also be passed as the `out` keyword argument.

**Changes to maintain backwards compatibility**
* Allow a single array or None to be passed as the `out` keyword argument (if no positional output arrays have been provided), even if `nout > 1`, and use it for the first array.

jaimefrio commented:
OK, this seems to be working now. I still need to document the new behavior, and since I also had to make changes to the array wrapping might be a good idea to add some tests for that. But I'd appreciate a first review.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -960,35 +976,69 @@ static int get_ufunc_arguments(PyUFuncObject *ufunc,
+                        }
+                        else if (nout == 1) {
+                            /* Can be an array if it only has one output */
+                            if (_check_out_array(value, out_op+nin) < 0) {
```
Overall, to the extent I can judge, this looks good and clearer than the code that was present. A perhaps nitpick is that I think assignments inside if functions make the code less clear than it could be otherwise. I would suggest replacing this stanza (and similar ones) with something like the following:
```
array_ok = _set_out_array(value, out_op+nin)
if (array_ok < 0) {
    ...
```
Here, I also changed the name to ensure it is clear what the function does; it could also be `_check_and_set_out_array` or `_set_out_array_if_not_None` or so).

jaimefrio replied:
Good suggestion on the name of the function, will change. But single function calls inside an if statement, when the return isn't needed elsewhere, are everywhere in the numpy code base, and in general in all Python C extensions, starting with the `if (!PyArg_ParseTupleAndKeywords...) {return NULL;}` in [the python.org tutorial on C extensions](https://docs.python.org/2/extending/extending.html#keyword-parameters-for-extension-functions).

I really think that adding a variable is unnecessary bulk in this case, so am going to stand my ground on that one, unless there is an overwhelming opinion against.

mhvk replied:
Not being used very much to `C` I'm happy to follow convention -- though this particular piece of code did not seem to follow it originally.  Anyway, I think that with the name change it will be much more obvious that the main action is something being *set*, not just *checked*.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -4043,45 +4125,41 @@ _find_array_wrap(PyObject *args, PyObject *kwds,
-            }
+
+    if (nin == nargs && kwds != NULL &&
+                        (obj = PyDict_GetItem(kwds, npy_um_str_out))) {
```
Here also I would suggest taking the assignment out of the `if` statement and then simply doing another `if`.

jaimefrio replied:
This makes more sense, I don't like it much either, although it makes the branching structure very compact. Will rethink it, see if I can come up with something better.

jaimefrio commented:
I have incorporated some of Marten's suggestions, and added some more tests to validate the wrapping behavior. Documenting the new behavior is the only thing left, but the code itself should be almost ready to go.

jaimefrio commented:
I have made some changes, to properly report the error when the tuple does not have as many items as output arguments. I have also added some testing for the proper errors being raised. I think now only the documentation part is left.

jaimefrio commented:
Ok, have added the documentation changes too. I think it is ready to go.

mhvk commented on doc/source/reference/ufuncs.rst:
```diff
@@ -299,7 +299,20 @@ advanced usage and will not typically be used.
+
+    ..versionadded:: 1.10
+
+    The 'out' keyword argument is now expected to be a tuple with one entry
```
I'd delete "now" here. Documentation should aim at those without previous knowledge; the `versionadded` will be enough for those who did know.

mhvk commented on numpy/core/tests/test_umath.py:
```diff
@@ -36,35 +36,137 @@ def test_e(self):
+            def __array_wrap__(self, arr, context):
+                return arr.view(type(self))
+
+        for b in (True, False):
```
How about renaming `b` to `subok`?

mhvk commented:
Looks good! Only two very minor comments.

jaimefrio commented:
Minor but sensible, should be corrected in the latest push.

charris commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -878,23 +910,7 @@ static int get_ufunc_arguments(PyUFuncObject *ufunc,
-            PyErr_SetString(PyExc_TypeError,
-                            "return arrays must be "
-                            "of ArrayType");
+        if (_set_out_array(obj, out_op+i) < 0) {
```
Space around `+`.

charris commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -960,35 +976,73 @@ static int get_ufunc_arguments(PyUFuncObject *ufunc,
+                    /*
+                     * Output arrays may be specified as a keyword argument,
+                     * either as a single array (or None) for single output
+                     * ufuncs, or as a tuple of arrays (or Nones).
```
I'd leave of the parenthesis around `or None`, and maybe `and Nones` instead.

charris commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -960,35 +976,73 @@ static int get_ufunc_arguments(PyUFuncObject *ufunc,
+                                        "one entry per ufunc output");
+                                goto fail;
+                            }
+                            /* 'out' must be a tuple of arrays */
```
and Nones.

charris commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -960,35 +976,73 @@ static int get_ufunc_arguments(PyUFuncObject *ufunc,
+                        }
+                        else if (nout == 1) {
+                            /* Can be an array if it only has one output */
+                            if (_set_out_array(value, out_op+nin) < 0) {
```
Spaces around `+`.

charris commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -960,35 +976,73 @@ static int get_ufunc_arguments(PyUFuncObject *ufunc,
+                             * keep only the else branch of this if-else
+                             */
+                            if (PyArray_Check(value) || value == Py_None) {
+                                if (DEPRECATE("passing a single array to the "
```
Might want to scatter a `\n` or two in the message.

charris commented:
Few style nitpicks, otherwise LGTM. Does this affect #5566?

jaimefrio commented:
Fixed.

About #5566, I don't think it affects it, since the code seems to already be considering the possibility of `out` being a tuple of arrays. Can you confirm, @cowlicks?

charris commented:
Thanks Jaime. Noticed a few more places needing spaces, but I think we should make a script to find those and do a general cleanup at some point.

juliantaylor commented:
thanks this is great, I think using tuple makes much more sense, makes me wonder why was this not originally done like this.

though now: ufunc(x, out=x, out2=y)` only raises an warning while previously it raised a value error:

```python
x = np.arange(9)
o1 = np.empty(9)
o2 = np.empty(9, dtype=np.int32)
np.frexp(x, out=o1, out2=o2)

ValueError: cannot specify 'out' as both a positional and keyword argument
```

jaimefrio commented:
Should we be more strict in the checking of `'out'` being just `'out'`?

njsmith commented:
> Should we be more strict in the checking of 'out' being just 'out'?

+10

jaimefrio commented:
I'm not sure why the code uses `strncmp` and not directly `strcmp`, which is all it would take to do proper checking. Am I missing something?

njsmith commented:
I guess it's either an ancient thinko, or else maybe someone was planning
to actually support out1=, out2=, like the docstrings say, but then got
distracted halfway through. Either way it doesn't much matter at this
point...
On Mar 9, 2015 4:58 PM, "Jaime" <notifications@github.com> wrote:

> I'm not sure why the code uses strncmp and not directly strcmp, which is
> all it would take to do proper checking. Am I missing something?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5621#issuecomment-77970188>.
>

mhvk commented:
:+1: to checking `out` properly; it definitely confused me the first time I encountered it.

jaimefrio commented:
Take a look at #5659 .
